File: serialportk/src/main/java/com/mozhimen/serialportk/SerialPortKFinder.java
/*
 * Copyright 2009 Cedric Priscal
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License. 
 */

package com.mozhimen.serialportk;

import android.util.Log;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.LineNumberReader;
import java.util.Iterator;
import java.util.Vector;

public class SerialPortKFinder {

    private static final String TAG = "SerialPortKFinder>>>>>";

    private Vector<SerialPortDriver> _serialPortDrivers = null;

    Vector<SerialPortDriver> getDrivers() throws IOException {
        if (_serialPortDrivers == null) {
            _serialPortDrivers = new Vector<>();
            LineNumberReader r = new LineNumberReader(new FileReader("/proc/tty/drivers"));
            String l;
            while ((l = r.readLine()) != null) {
                // Issue 3:
                // Since driver name may contain spaces, we do not extract driver name with split()
                String driverName = l.substring(0, 0x15).trim();
                String[] w = l.split(" +");
                if ((w.length >= 5) && (w[w.length - 1].equals("serial"))) {
                    Log.d(TAG, "Found new driver " + driverName + " on " + w[w.length - 4]);
                    _serialPortDrivers.add(new SerialPortDriver(driverName, w[w.length - 4]));
                }
            }
            r.close();
        }
        return _serialPortDrivers;
    }

    public String[] getAllDevices() {
        Vector<String> devices = new Vector<>();
        // Parse each driver
        Iterator<SerialPortDriver> iteratorDriver;
        try {
            iteratorDriver = getDrivers().iterator();
            while (iteratorDriver.hasNext()) {
                SerialPortDriver serialPortDriver = iteratorDriver.next();
                Iterator<File> iteratorDevices = serialPortDriver.getDevices().iterator();
                while (iteratorDevices.hasNext()) {
                    String device = iteratorDevices.next().getName();
                    String value = String.format("%s (%s)", device, serialPortDriver.getName());
                    devices.add(value);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return devices.toArray(new String[devices.size()]);
    }

    public String[] getAllDevicesPath() {
        Vector<String> devices = new Vector<>();
        // Parse each driver
        Iterator<SerialPortDriver> iteratorDriver;
        try {
            iteratorDriver = getDrivers().iterator();
            while (iteratorDriver.hasNext()) {
                SerialPortDriver serialPortDriver = iteratorDriver.next();
                Iterator<File> iteratorDevice = serialPortDriver.getDevices().iterator();
                while (iteratorDevice.hasNext()) {
                    String device = iteratorDevice.next().getAbsolutePath();
                    devices.add(device);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return devices.toArray(new String[devices.size()]);
    }

    public class SerialPortDriver {
        private String _driverName;
        private String _deviceRoot;
        Vector<File> _devices = null;

        public SerialPortDriver(String name, String root) {
            _driverName = name;
            _deviceRoot = root;
        }

        public Vector<File> getDevices() {
            if (_devices == null) {
                _devices = new Vector<>();
                File dev = new File("/dev");

                File[] files = dev.listFiles();

                if (files != null) {
                    int i;
                    for (i = 0; i < files.length; i++) {
                        if (files[i].getAbsolutePath().startsWith(_deviceRoot)) {
                            Log.d(TAG, "Found new device: " + files[i]);
                            _devices.add(files[i]);
                        }
                    }
                }
            }
            return _devices;
        }

        public String getName() {
            return _driverName;
        }

        @Override
        public String toString() {
            return "SerialPortDriver{" +
                    "_driverName='" + _driverName + '\'' +
                    ", _deviceRoot='" + _deviceRoot + '\'' +
                    ", _devices=" + _devices +
                    '}';
        }
    }
}

File: serialportk/src/main/java/com/mozhimen/serialportk/SerialPortKFinder.kt
package com.mozhimen.serialportk

import android.util.Log
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.io.LineNumberReader
import java.util.*


/**
 * @ClassName SerialPortKFinder
 * @Description TODO
 * @Author Mozhimen & Kolin Zhao
 * @Date 2022/12/5 15:32
 * @Version 1.0
 */
class SerialPortKFinder {
    companion object {
        private const val TAG = "SerialPortKFinder>>>>>"
    }

    private var _drivers: Vector<SerialPortKDriver>? = null

    @Throws(IOException::class)
    fun getDrivers(): Vector<SerialPortKDriver> {
        if (_drivers == null) {
            _drivers = Vector()
            val r = LineNumberReader(FileReader("/proc/tty/drivers"))
            var l: String
            while (r.readLine().also { l = it } != null) {
                // Issue 3:
                // Since driver name may contain spaces, we do not extract driver name with split()
                val driverName = l.substring(0, 0x15).trim { it <= ' ' }
                val w = l.split(" +").toTypedArray()
                if (w.size >= 5 && w[w.size - 1] == "serial") {
                    Log.d(TAG, "Found new driver " + driverName + " on " + w[w.size - 4])
                    _drivers!!.add(SerialPortKDriver(driverName, w[w.size - 4]))
                }
            }
            r.close()
        }
        return _drivers!!
    }

    fun getAllDevices(): Array<String> {
        val devices = Vector<String>()
        // Parse each driver
        val iteratorDriver: Iterator<SerialPortKDriver>
        try {
            iteratorDriver = getDrivers().iterator()
            while (iteratorDriver.hasNext()) {
                val driver = iteratorDriver.next()
                val iteratorDevice: Iterator<File> = driver.devices.iterator()
                while (iteratorDevice.hasNext()) {
                    val device = iteratorDevice.next().name
                    devices.add("$device (${driver.name})")
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return devices.toTypedArray()
    }

    fun getAllDevicesPath(): Array<String> {
        val devices = Vector<String>()
        // Parse each driver
        val iteratorDriver: Iterator<SerialPortKDriver>
        try {
            iteratorDriver = getDrivers().iterator()
            while (iteratorDriver.hasNext()) {
                val driver = iteratorDriver.next()
                val iteratorDevice: Iterator<File> = driver.devices.iterator()
                while (iteratorDevice.hasNext()) {
                    val device = iteratorDevice.next().absolutePath
                    devices.add(device)
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return devices.toTypedArray()
    }
}